unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ExploreHUD/Scripts/Friends/ExploreFriendsController.cs

```diff
@@ -0,0 +1,263 @@
+                    friendIterator.Current.Value.AddListener(listener);
+                }
+            }
+        }
+    }
+}
+
+class FriendWrapper
+{
+    HashSet<ListenerWrapper> friendListeners = new HashSet<ListenerWrapper>();
+
+    public UserProfile profile { private set; get; }
+    public FriendsController.UserStatus status { private set; get; }
+
+    public FriendWrapper(string userId)
+    {
+        profile = UserProfileController.userProfilesCatalog.Get(userId);
+    }
+
+    public void SetStatus(FriendsController.UserStatus newStatus)
+    {
+        status = newStatus;
+    }
+
+    public void AddListener(ListenerWrapper listener)
+    {
+        listener.OnListenerDisposed += OnListenerDisposed;
+        friendListeners.Add(listener);
+        listener.OnFriendAdded(profile);
+    }
+
+    public void RemoveListener(ListenerWrapper listener)
+    {
+        OnListenerRemoved(listener);
+        friendListeners.Remove(listener);
+    }
+
+    public bool HasListeners()
+    {
+        return friendListeners.Count > 0;
+    }
+
+    public bool HasChangedLocation(Vector2Int coords)
+    {
+        if (friendListeners.Count > 0 && friendListeners.First().ContainCoords(coords))
+        {
+            return false;
+        }
+        return true;
+    }
+
+    public void RemoveAllListeners()
+    {
+        if (!HasListeners())
+        {
+            return;
+        }
+
+        using (var listenerIterator = friendListeners.GetEnumerator())
+        {
+            while (listenerIterator.MoveNext())
+            {
+                OnListenerRemoved(listenerIterator.Current);
+            }
+        }
+        friendListeners.Clear();
+    }
+
+    public bool IsOnline()
+    {
+        if (status.presence != PresenceStatus.ONLINE)
+            return false;
+        if (status.realm == null)
+            return false;
+        return !string.IsNullOrEmpty(status.realm.serverName) && !string.IsNullOrEmpty(status.realm.layer);
+    }
+
+    void OnListenerDisposed(ListenerWrapper listener)
+    {
+        listener.OnListenerDisposed -= OnListenerDisposed;
+        friendListeners.Remove(listener);
+    }
+
+    void OnListenerRemoved(ListenerWrapper listener)
+    {
+        listener.OnListenerDisposed -= OnListenerDisposed;
+        listener.OnFriendRemoved(profile);
+    }
+}
+
+class ListenerWrapper : IDisposable
+{
+    IExploreViewWithFriends view;
+
+    public event Action<ListenerWrapper> OnListenerDisposed;
+
+    public ListenerWrapper(IExploreViewWithFriends view)
+    {
+        this.view = view;
+    }
+
+    public void OnFriendAdded(UserProfile profile)
+    {
+        view.OnFriendAdded(profile);
+    }
+
+    public void OnFriendRemoved(UserProfile profile)
+    {
+        view.OnFriendRemoved(profile);
+    }
+
+    public bool ContainCoords(Vector2Int coords)
+    {
+        return view.ContainCoords(coords);
+    }
+
+    public void Dispose()
+    {
+        OnListenerDisposed?.Invoke(this);
+    }
+}
```

Comment:
The existence of these wrapper classes is really justified? What are the responsibilities here?

Reply:
`ListenerWrapper` is created when a Scene cell is added.
It mostly to handle when a Scene cell is removed (so it doesn't need to listen to friends state anymore) to tell the `FriendWrapper` (friend that is in some of the Scene cell coords) to remove it from it listeners list.
An option to remove this wrapper is iterating through all friends and check every listener in their list to remove them.

`FriendWrapper` contains a list of `ListenerWrapper`. At this moment the list only consist in only one element, but taking in mind what is planned for this "explore" feature in the feature, It might be possible that more than one cell will need to listen to the same friend status